File: leetcode/2326.spiral-matrix-iv/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func spiralMatrix(m int, n int, head *ListNode) [][]int {
	ans := make([][]int, m)
	for i := range m {
		ans[i] = make([]int, n)
		for j := range n {
			ans[i][j] = -1
		}
	}

	row := 0
	col := 0
	dirs := [][]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
	d := 0

	for head != nil {
		ans[row][col] = head.Val

		if dr, dc := dirs[d][0], dirs[d][1]; row+dr >= m || col+dc >= n || col+dc < 0 || ans[row+dr][col+dc] != -1 {
			d = (d + 1) % 4
		}

		row += dirs[d][0]
		col += dirs[d][1]

		head = head.Next
	}

	return ans
}

func main() {
	arr := []int{3, 0, 2, 6, 8, 1, 7, 9, 4, 2, 5, 5, 0}
	const m = 3
	const n = 5

	head := ListNode{arr[0], nil}
	tmp := &head

	for i := 1; i < len(arr); i++ {
		newNode := ListNode{arr[i], nil}
		tmp.Next = &newNode
		tmp = &newNode
	}

	ans := spiralMatrix(m, n, &head)

	fmt.Println(ans)
}

File: leetcode/2326.spiral-matrix-iv/solution.c
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

int **spiralMatrix(int m, int n, struct ListNode *head, int *returnSize,
                   int **returnColumnSizes)
{

    int **res = (int **)malloc(m * sizeof(int *));
    *returnSize = m;
    *returnColumnSizes = (int *)malloc(m * sizeof(int));

    for (int i = 0; i < m; ++i) {
        res[i] = (int *)malloc(n * sizeof(int));
        (*returnColumnSizes)[i] = n;

        for (int j = 0; j < n; ++j) {
            res[i][j] = -1;
        }
    }

    int r = 0;
    int c = 0;
    int dr = 0;
    int dc = 1;

    while (head) {
        res[r][c] = head->val;
        if (r + dr == m || c + dc == n || c + dc < 0 ||
            res[r + dr][c + dc] != -1) {
            // change direction clockwise
            int tmp = dr;
            dr = dc;
            dc = -tmp;
        }
        r += dr;
        c += dc;
        head = head->next;
    }

    return res;
}

int main(void) { return EXIT_SUCCESS; }
